<launch>

	<arg name="use_sim_time" default="false"/>
	<group if="$(arg use_sim_time)">
		<param name="/use_sim_time" value="true" />
	</group>

	<arg name="map_num" default="$(optenv GM_PARAM_MAP_NUM 0)"/>

	<arg name="stdr_map" default="$(find dynamic_gap)/maps/campus_maps/$(arg map_num).yaml"/>
	<arg name="map_file" default="$(find dynamic_gap)/maps/campus.yaml"/>
	
	<node name="known_map" pkg="map_server" type="map_server" args="$(arg map_file)" required="true">
		<remap from="/map" to="/known_map"/>
		<remap from="/map_metadata" to="/known_map_metadata"/>
		<param name="frame_id" value="known_map"/>
	</node>

	<node pkg="tf" type="static_transform_publisher" name="world2kmap" args="0 0 0 0 0 0  world known_map 20" />

	<!-- need to re-call this? -->
	<node type="stdr_server_node" pkg="stdr_server" name="stdr_server" output="screen" args="$(arg stdr_map)"/>

	<node pkg="tf" type="static_transform_publisher" name="world2map" args="0 0 0 0 0 0  world map 20" />
	
	<include file="$(find stdr_gui)/launch/stdr_gui.launch"/>

        <include file="$(find stdr_robot)/launch/robot_manager.launch" />

        <node pkg="stdr_robot" type="robot_handler" name="robot0" args="add $(find dynamic_gap)/stdr_robots/robots/holonomic_robot_360_bumper.xml 1 1 0" />

	<node pkg="amcl" type="amcl" name="amcl">

	    <param name="min_particles"             value="500"/>
	    <param name="max_particles"             value="3000"/>
	    <param name="kld_err"                   value="0.02"/>
	    <param name="update_min_d"              value="0.20"/>
	    <param name="update_min_a"              value="0.20"/>
	    <param name="resample_interval"         value="1"/>
	    <param name="transform_tolerance"       value="0.5"/>
	    <param name="recovery_alpha_slow"       value="0.00"/>
	    <param name="recovery_alpha_fast"       value="0.00"/>
	    <param name="initial_pose_x"            value="0.00"/>
	    <param name="initial_pose_y"            value="0.00"/>
	    <param name="initial_pose_a"            value="0.00"/>
	    <param name="gui_publish_rate"          value="50.0"/>

	    <remap from="scan"                      to="robot0/laser_0"/>
	    <param name="laser_max_range"           value="3.5"/>
	    <param name="laser_max_beams"           value="180"/>
	    <param name="laser_z_hit"               value="0.5"/>
	    <param name="laser_z_short"             value="0.05"/>
	    <param name="laser_z_max"               value="0.05"/>
	    <param name="laser_z_rand"              value="0.5"/>
	    <param name="laser_sigma_hit"           value="0.2"/>
	    <param name="laser_lambda_short"        value="0.1"/>
	    <param name="laser_likelihood_max_dist" value="2.0"/>
	    <param name="laser_model_type"          value="likelihood_field"/>

	    <param name="odom_model_type"           value="diff"/>
	    <param name="odom_alpha1"               value="0.1"/>
	    <param name="odom_alpha2"               value="0.1"/>
	    <param name="odom_alpha3"               value="0.1"/>
	    <param name="odom_alpha4"               value="0.1"/>
	    <param name="odom_frame_id"             value="map_static"/>
	    <param name="base_frame_id"             value="robot0"/>
            <param name="global_frame_id"           value="known_map"/>

	</node>

  <node pkg="move_base" type="move_base" respawn="false" name="move_base">
      <!-- Default configs form the TB3 repos -->
      <!-- <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" /> -->
      <rosparam file="$(find turtlebot3_navigation)/param/costmap_common_params_burger.yaml" command="load" ns="global_costmap" />
      <rosparam file="$(find turtlebot3_navigation)/param/costmap_common_params_burger.yaml" command="load" ns="local_costmap" />
      <rosparam file="$(find turtlebot3_navigation)/param/local_costmap_params.yaml" command="load" />
      <rosparam file="$(find turtlebot3_navigation)/param/global_costmap_params.yaml" command="load" />
      <rosparam file="$(find turtlebot3_navigation)/param/move_base_params.yaml" command="load" />
      <rosparam file="$(find turtlebot3_navigation)/param/dwa_local_planner_params_burger.yaml" command="load" />

      <!-- Set tf_prefix for frames explicity, overwriting defaults -->
      <param name="global_costmap/scan/sensor_frame" value="robot0/laser_0"/>
      <param name="global_costmap/scan/topic" value="robot0/laser_0"/>
      <param name="global_costmap/obstacle_layer/scan/sensor_frame" value="robot0/base_scan"/>
      <param name="global_costmap/global_frame" value="known_map"/>
      <param name="global_costmap/robot_base_frame" value="robot0"/>

      <param name="local_costmap/scan/sensor_frame" value="robot0/laser_0"/>
      <!-- <param name="local_costmap/scan/topic" value="robot0/laser_0"/> -->
      <param name="local_costmap/obstacle_layer/scan/sensor_frame" value="robot0/base_scan"/>
      <param name="local_costmap/global_frame" value="known_map"/>
      <param name="local_costmap/robot_base_frame" value="robot0"/>

      <!-- Centralized map server -->
      <remap from="map" to="/map"/>
  </node>
</launch>
